include/inexor/vulkan-renderer/world/collision.hpp

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include <glm/vec3.hpp>
+
+#include <string>
+#include <tuple>
+
+namespace inexor::vulkan_renderer::world {
+
+/// @brief A wrapper for collisions between a ray and octree geometry.
+/// This class is used for octree collision, but it can be used for every cube-like data structure
+/// @tparam T A templatable type which offers a size() and center() method.
+template <typename T>
+class RayCubeCollision {
+    const T &m_cube;
+
+    using cube_face = std::tuple<std::string, glm::vec3>;
+    using cube_corner = std::tuple<std::string, glm::vec3>;
+    using cube_edge = std::tuple<std::string, glm::vec3>;
+
+    glm::vec3 m_intersection;
+    cube_face m_selected_face;
+    cube_corner m_nearest_corner;
+    cube_edge m_nearest_edge;
+
+public:
+    /// @brief Calculate point of intersection, selected face,
+    /// nearest corner on that face, and nearest edge on that face.
+    /// @param cube The cube to check for collision.
+    /// @param ray_pos The start point of the ray.
+    /// @param ray_dir The direction of the ray.
+    RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir);
```

Comment:
In this case, we should even consider `std::weak_ptr`.

Reply:
As a sidenote: this would be a good scenario for atomic smart pointers. Maybe we will have them one day.

Reply:
I will keep it as a const reference for now, since we need to rethink the entire engine in terms of const references when it comes to thread safety anyways.